extensions/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -0,0 +1,925 @@
+   description is fed to *clCreateProgramWithDefinedBuiltinKernels*
+   call, then implementation must return
+   `CL_DBK_UNSUPPORTED_PROPERTY_KHR`.
+
+[[dbk-propery-table]]
+.Table of defined built-in kernel properties
+[cols="2,1,2",stripes=odd]
+|===
+| *DBK Mode Property* | *Property Value* | *Description*
+
+| CL_DBK_PROPERTY_MAX_RELATIVE_ERROR_KHR | float
+
+a| Require that the DBK produces the results which do not deviate more
+than the given amount value of ULPs (units in the last place) respect
+to infnitely precise result.
+
+| CL_DBK_PROPERTY_NON_DETERMINISTIC_KHR | cl_bool
+
+a| Allow results of the kernel to be non-reproducible. This allows
+implementation to switch algorithm of the kernel on each launch for
+possibly better performance.
+// Idea from https://pytorch.org/docs/stable/notes/randomness.html#cuda-convolution-benchmarking
+
+|===
+
+
+[[dbk-description-table]]
+.Standard Built-in Kernels and Their Semantics. *The table has been populated with a small set of non-trivial example entries which are subject to change and the list to expand during drafting.*
+|===
+| Name: *CL_DBK_GEMM_KHR*
+| *Kernel Attributes*
+a|
+
+[source,c]
+----
+typedef struct cl_dbk_attributes_gemm_khr {
+    cl_tensor_desc a;
+    cl_tensor_desc b;
+    cl_tensor_desc c_in;
+    cl_tensor_desc c_out;
+    cl_bool trans_a;
+    cl_bool trans_b;
+    cl_tensor_datatype_union_khr alpha;
+    cl_tensor_datatype_union_khr beta;
+    cl_dbk_properties kernel_props[CL_MAX_DBK_PROPERTIES];
+} cl_dbk_attributes_gemm_khr;
+----
+
+* _a_ is a tensor description for input matrix A.
+
+* _b_ is a tensor description for input matrix B.
+
+* _c_in_ is a tensor description for output matrix CIN.
+
+* _c_out_ is a tensor description for output matrix COUT.
+
+* _trans_a_ instruct to transpose the A matrix if the value is CL_TRUE.
+
+* _trans_b_ instruct to transpose the B matrix if the value is CL_TRUE.
+
+* _alpha_ is a value or pointer to value corresponponding to the
+  element type of _c_out_.
+
+* _beta_ is a value or pointer to value corresponponding to the
+  element type of _c_out_.
+
+* _kernel_props_ defined additional kernel properties.
+
+| *Kernel Arguments*
+a|
+. cl_mem: a tensor object for matrix A (read only).
+. cl_mem: a tensor object for matrix B (read only).
+. cl_mem: a tensor object for matrix C_IN (read only).
+. cl_mem: a tensor object for matrix C_OUT (write only).
+
+| *Description* a| Performs (batched) general matrix multiplication:
+
+[stem]
+++++
+bb"COUT"_(b,m,n) = "beta" * bb"CIN"_(b,m,n) + "alpha" * sum_(k)trans(bb"A", "trans_a")_(b,m,k)trans(bb"B", "trans_b") _(b,k,n)
+++++
+
+Where:
+
+[stem]
+++++
+trans(X_(b,i,j), tr) = {(X_(b,j,i), "if tr" = "CL_TRUE"), (X_(b,i,j), "otherwise") :}
+++++
+
+Second degree tensors of shape `(a, b)` are treated as third degree
+tensors of shape `(1, a, b)`.
+
+Operations of the matrix muliplication are performed in the precision
+of the `elementof\(COUT)`.
```

Comment:
There seem to be a couple of formatting errors regarding `\(COUT)`

Reply:
Thanks for pointing this out. It'll be fixed.